drivers/tee/optee/optee.c

```diff
@@ -205,6 +228,80 @@ static inline bool check_param_input(struct optee_msg_arg *arg)
 
 static void *optee_construct_page_list(void *buf, uint32_t len, uint64_t *phys_buf);
 
+static uint32_t optee_call_supp(const struct device *dev, uint32_t func, size_t num_params,
+			struct tee_param *param)
+
+{
+	struct optee_driver_data *data = (struct optee_driver_data *) dev->data;
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req;
+	uint32_t ret;
+
+	req = k_malloc(sizeof(*req));
+	if (!req) {
+		return TEEC_ERROR_OUT_OF_MEMORY;
+	}
+
+	k_sem_init(&req->complete, 0, 1);
+	req->func = func;
+	req->num_params = num_params;
+	req->param = param;
+
+	/* Insert the request in the request list */
+	k_mutex_lock(&supp->mutex, K_FOREVER);
+	sys_dlist_append(&supp->reqs, &req->link);
+	k_mutex_unlock(&supp->mutex);
+
+	/* Tell an eventual waiter there's a new request */
+	k_sem_give(&supp->reqs_c);
+
+	/*
+	 * Wait for supplicant to process and return result, once we've
+	 * returned from k_sem_take(&req->c) successfully we have
+	 * exclusive access again.
+	 */
+
+	k_sem_take(&req->complete, K_FOREVER);
+
+	ret = req->ret;
+	k_free(req);
+
+	return ret;
+}
+
+static int cmd_alloc_suppl(const struct device *dev, size_t sz, struct tee_shm **shm)
+{
+	uint32_t ret;
+	struct tee_param param;
+
+	param.attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT;
+	param.a = OPTEE_RPC_SHM_TYPE_APPL;
+	param.b = sz;
+	param.c = 0;
+
+	ret = optee_call_supp(dev, OPTEE_RPC_CMD_SHM_ALLOC, 1, &param);
+
+	if (ret) {
+		return ret;
+	}
+
+	ret = tee_add_shm(dev, (void *)param.c, 0, param.b, 0, shm);
+
+	return ret;
+}
+
+static void cmd_free_suppl(const struct device *dev, struct tee_shm *shm)
+{
+	struct tee_param param;
+
+	param.attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT;
+	param.a = OPTEE_RPC_SHM_TYPE_APPL;
+	param.b = (uint64_t)shm;
+	param.c = 0;
+
+	optee_call_supp(dev, OPTEE_RPC_CMD_SHM_FREE, 1, &param);
```

Comment:
```suggestion
	optee_call_supp(dev, OPTEE_RPC_CMD_SHM_FREE, 1, &param);
	tee_rm_shm(dev, shm);
```

we must free `shm` allocated in `cmd_alloc_suppl()`